File: projects/06/assemblers/Kotlin/src/main/kotlin/Assembler.kt
package main.kotlin

class Assembler {
    fun assemble(program: String): String {
        val parser = Parser(program)
        val symTable = SymbolTable()
        val code = Code()

        // First pass to record symbols
        var currentLine = 0
        while (parser.hasMoreCommands()) {
            val commandType: CommandType = parser.commandType()

            if (commandType == CommandType.L_COMMAND) {
                symTable.addEntry(parser.symbol(), currentLine)
            } else {
                currentLine++
            }

            parser.advance()
        }

        parser.reset()
        val assemblyCode = StringBuilder()

        while (parser.hasMoreCommands()) {
            val commandType: CommandType = parser.commandType()

            if (commandType == CommandType.C_COMMAND) {
                assemblyCode.append(code.genCommand(parser.parseCommand()))
                assemblyCode.append("\n")
            } else if (commandType == CommandType.A_COMMAND) {
                val sym = parser.symbol()
                val address = when {
                    (sym.toIntOrNull() != null) -> sym.toInt()
                    (symTable.contains(sym)) -> symTable.GetAddress(sym)
                    else -> {
                        symTable.addEntry(sym)
                        symTable.GetAddress(sym)
                    }
                }
                assemblyCode.append(address?.toString(2)?.padStart(16, '0'))
                assemblyCode.append("\n")
            }

            parser.advance()
        }

        return assemblyCode.toString()
    }
}

File: projects/06/assemblers/rust/src/assembler.rs
use crate::code_gen::generate_command;
use crate::parser::{CommandType, Parser};
use crate::symbols::Symbols;

pub struct Assembler {}

impl Assembler {
    pub fn assemble(raw_program: String) -> String {
        let mut parser = Parser::new(&raw_program);
        let mut symbol_table = Symbols::new();

        // first pass to record jump symbols
        let mut line_index = 0;
        while parser.has_more_commands() {
            match parser.command_type() {
                CommandType::Jump => {
                    symbol_table.add_jump_symbol(parser.get_symbol(), line_index);
                }
                _ => {
                    line_index += 1;
                }
            }
            parser.advance();
        }

        parser.reset();

        // start assembly
        let mut assembly_code = String::new();
        while parser.has_more_commands() {
            match parser.command_type() {
                CommandType::Address => {
                    let symbol = parser.get_symbol();
                    let address = symbol
                        .parse::<u16>()
                        .unwrap_or_else(|_| symbol_table.get_address(symbol));

                    let binary_str = format!("{:0>16b}", address);
                    assembly_code.push_str(&binary_str);
                    assembly_code.push_str("\n");
                }
                CommandType::Computation => {
                    let command = parser.parse_command();
                    assembly_code.push_str(&generate_command(command));
                    assembly_code.push_str("\n");
                }
                _ => {}
            }

            parser.advance();
        }

        assembly_code
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn assemble() {
        let program = "
           // Test Program
           
           @42
           D=M;JLT // will never jump
           
           @foo // not used
           @bar
           M=D
           
           (END)
           @END
           0;JMP
           
        "
        .to_string();

        let mut output = String::new();
        output.push_str("0000000000101010\n");
        output.push_str("1111110000010100\n");
        output.push_str("0000000000010000\n");
        output.push_str("0000000000010001\n");
        output.push_str("1110001100001000\n");
        output.push_str("0000000000000101\n");
        output.push_str("1110101010000111\n");

        assert_eq!(Assembler::assemble(program), output);
    }
}
